File: 00215. Kth Largest Element in an Array/00215.cpp
// https://leetcode.com/problems/kth-largest-element-in-an-array/

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> pq(nums.begin(), nums.end());
        while (--k > 0) {
            pq.pop();
        }
        return pq.top();
    }
};

File: LeetCode/00215. Kth Largest Element in an Array/00215.go
// https://leetcode.com/problems/kth-largest-element-in-an-array

type IntHeap []int

func (h IntHeap) Len() int {
    return len(h)
}

// Min-Heap
func (h IntHeap) Less(i, j int) bool {
    return h[i] < h[j]
}

func (h IntHeap) Swap(i, j int) {
    h[i], h[j] = h[j], h[i]
}

func (h *IntHeap) Push(x any) {
    *h = append(*h, x.(int))
}

func (h *IntHeap) Pop() any {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[:n-1]
    return x
}

func findKthLargest(nums []int, k int) int {
    h := &IntHeap{}
    heap.Init(h)

    for _, n := range nums {
        heap.Push(h, n)

        if h.Len() > k {
            heap.Pop(h)
        }
    }

    return (*h)[0]
}
